File: apps/server/agent/src/systemprompt/walk.ts
/**
 * Filesystem walk-up helpers for config discovery.
 * Mirrors oh-my-pi discovery ancestor walks without the capability registry.
 */
import * as fs from "node:fs/promises";
import * as os from "node:os";
import * as path from "node:path";
import type { ConfigLevel, SourceMeta } from "../types/system-prompt.js";

/** Project-level config directory names (relative to each ancestor). */
export const PROJECT_CONFIG_DIRS = [".console", ".agent", ".agents"] as const;

/** User-level config directory names under home. */
export const USER_CONFIG_DIRS = [".console", ".agent", ".agents"] as const;

export interface DiscoveryRoots {
  cwd: string;
  home: string;
  /** Optional stop path for project walk (repo root). */
  stopAt?: string;
}

export function resolveRoots(options: {
  cwd?: string;
  home?: string;
  stopAt?: string;
}): DiscoveryRoots {
  return {
    cwd: path.resolve(options.cwd ?? process.cwd()),
    home: options.home ?? os.homedir(),
    stopAt: options.stopAt ? path.resolve(options.stopAt) : undefined,
  };
}

export function createSourceMeta(
  provider: string,
  filePath: string,
  level: ConfigLevel,
): SourceMeta {
  return {
    provider,
    path: path.resolve(filePath),
    level,
  };
}

/** Ancestors from cwd up to stopAt / filesystem root (cwd first). */
export function getAncestorDirs(
  cwd: string,
  stopAt?: string,
): Array<{ dir: string; depth: number }> {
  const ancestors: Array<{ dir: string; depth: number }> = [];
  let current = path.resolve(cwd);
  let depth = 0;
  while (true) {
    ancestors.push({ dir: current, depth });
    if (stopAt && current === path.resolve(stopAt)) break;
    const parent = path.dirname(current);
    if (parent === current) break;
    current = parent;
    depth++;
  }
  return ancestors;
}

export async function readTextFile(filePath: string): Promise<string | null> {
  try {
    return await fs.readFile(filePath, "utf8");
  } catch (err: unknown) {
    const code = (err as NodeJS.ErrnoException).code;
    if (code === "ENOENT" || code === "ENOTDIR") return null;
    throw err;
  }
}

export async function pathExists(filePath: string): Promise<boolean> {
  try {
    await fs.access(filePath);
    return true;
  } catch {
    return false;
  }
}

export async function isDirectory(dirPath: string): Promise<boolean> {
  try {
    const stat = await fs.stat(dirPath);
    return stat.isDirectory();
  } catch {
    return false;
  }
}

/** List markdown-like files in a directory (non-recursive). */
export async function listMarkdownFiles(dirPath: string): Promise<string[]> {
  if (!(await isDirectory(dirPath))) return [];
  const entries = await fs.readdir(dirPath, { withFileTypes: true });
  return entries
    .filter(
      (e) =>
        e.isFile() &&
        (e.name.endsWith(".md") || e.name.endsWith(".mdc") || e.name.endsWith(".markdown")),
    )
    .map((e) => path.join(dirPath, e.name))
    .sort();
}

/** User config base dirs that exist under home. */
export async function getUserConfigDirs(home: string): Promise<string[]> {
  const dirs: string[] = [];
  for (const name of USER_CONFIG_DIRS) {
    const dir = path.join(home, name);
    if (await isDirectory(dir)) dirs.push(dir);
  }
  return dirs;
}

/**
 * Project config dirs for each ancestor: `<ancestor>/.console`, `.agent`, `.agents`.
 * Closest to cwd first.
 */
export async function getProjectConfigDirs(
  cwd: string,
  stopAt?: string,
): Promise<Array<{ dir: string; depth: number }>> {
  const result: Array<{ dir: string; depth: number }> = [];
  for (const { dir: ancestor, depth } of getAncestorDirs(cwd, stopAt)) {
    for (const name of PROJECT_CONFIG_DIRS) {
      const configDir = path.join(ancestor, name);
      if (await isDirectory(configDir)) {
        result.push({ dir: configDir, depth });
      }
    }
  }
  return result;
}

export function depthBetween(from: string, to: string): number {
  const fromParts = path.resolve(from).split(path.sep).filter(Boolean);
  const toParts = path.resolve(to).split(path.sep).filter(Boolean);
  // if `to` is ancestor of `from`, depth is difference in segments
  if (toParts.length > fromParts.length) return Math.abs(toParts.length - fromParts.length);
  let i = 0;
  while (i < toParts.length && i < fromParts.length && toParts[i] === fromParts[i]) i++;
  return fromParts.length - i;
}

File: apps/server-go/internal/agent/systemprompt/walk.go
// Filesystem walk-up helpers for config discovery. Port of
// apps/server/agent/src/systemprompt/walk.ts.
package systemprompt

import (
	"os"
	"path/filepath"
)

// ProjectConfigDirNames are project-level config directory names checked at
// each ancestor of cwd.
var ProjectConfigDirNames = []string{".console", ".agent", ".agents"}

// UserConfigDirNames are user-level config directory names checked under home.
var UserConfigDirNames = []string{".console", ".agent", ".agents"}

// ConfigLevel distinguishes user-home config from project-tree config.
type ConfigLevel string

const (
	LevelUser    ConfigLevel = "user"
	LevelProject ConfigLevel = "project"
)

// SourceMeta records where a discovered item was loaded from.
type SourceMeta struct {
	Provider string      `json:"provider"`
	Path     string      `json:"path"`
	Level    ConfigLevel `json:"level"`
}

func newSourceMeta(provider, path string, level ConfigLevel) SourceMeta {
	abs, err := filepath.Abs(path)
	if err != nil {
		abs = path
	}
	return SourceMeta{Provider: provider, Path: abs, Level: level}
}

// Roots resolves the effective cwd/home/stopAt for one discovery run.
type Roots struct {
	Cwd    string
	Home   string
	StopAt string
}

// ResolveRoots fills in defaults (process cwd, os home dir) for unset fields.
func ResolveRoots(cwd, home, stopAt string) Roots {
	if cwd == "" {
		if wd, err := os.Getwd(); err == nil {
			cwd = wd
		}
	}
	if abs, err := filepath.Abs(cwd); err == nil {
		cwd = abs
	}
	if home == "" {
		home, _ = os.UserHomeDir()
	}
	if stopAt != "" {
		if abs, err := filepath.Abs(stopAt); err == nil {
			stopAt = abs
		}
	}
	return Roots{Cwd: cwd, Home: home, StopAt: stopAt}
}

// AncestorDir is one directory on the walk from cwd up to the filesystem
// root (or StopAt), with its distance from cwd.
type AncestorDir struct {
	Dir   string
	Depth int
}

// AncestorDirs lists cwd and each parent up to stopAt (inclusive) or the
// filesystem root, cwd first.
func AncestorDirs(cwd, stopAt string) []AncestorDir {
	current := cwd
	if abs, err := filepath.Abs(current); err == nil {
		current = abs
	}
	var out []AncestorDir
	depth := 0
	for {
		out = append(out, AncestorDir{Dir: current, Depth: depth})
		if stopAt != "" && current == stopAt {
			break
		}
		parent := filepath.Dir(current)
		if parent == current {
			break
		}
		current = parent
		depth++
	}
	return out
}

func readTextFile(path string) (string, bool) {
	data, err := os.ReadFile(path)
	if err != nil {
		return "", false
	}
	return string(data), true
}

func isDir(path string) bool {
	info, err := os.Stat(path)
	return err == nil && info.IsDir()
}

// listMarkdownFiles lists .md/.mdc/.markdown files directly in dir, sorted.
func listMarkdownFiles(dir string) []string {
	if !isDir(dir) {
		return nil
	}
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil
	}
	var out []string
	for _, e := range entries {
		if e.IsDir() || !isMarkdownName(e.Name()) {
			continue
		}
		out = append(out, filepath.Join(dir, e.Name()))
	}
	return out
}

func isMarkdownName(name string) bool {
	for _, suffix := range []string{".md", ".mdc", ".markdown"} {
		if len(name) > len(suffix) && name[len(name)-len(suffix):] == suffix {
			return true
		}
	}
	return false
}

// UserConfigDirs returns existing user-level config dirs under home.
func UserConfigDirs(home string) []string {
	var out []string
	for _, name := range UserConfigDirNames {
		dir := filepath.Join(home, name)
		if isDir(dir) {
			out = append(out, dir)
		}
	}
	return out
}

// ProjectConfigDirs returns existing project-level config dirs at each
// ancestor of cwd (closest first), paired with their depth.
func ProjectConfigDirs(cwd, stopAt string) []AncestorDir {
	var out []AncestorDir
	for _, a := range AncestorDirs(cwd, stopAt) {
		for _, name := range ProjectConfigDirNames {
			dir := filepath.Join(a.Dir, name)
			if isDir(dir) {
				out = append(out, AncestorDir{Dir: dir, Depth: a.Depth})
			}
		}
	}
	return out
}
